ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc

```diff
@@ -118,15 +119,174 @@ int64 GetTransferredBytes(const NT& node) {
+  TF_RETURN_IF_ERROR(store.PutContexts(request, &response));
+  // Sets the node id for indicating this is an update when calling the APIs.
+  node.set_id(response.context_ids(0));
+  return tensorflow::Status::OK();
+}
+
+// Prepares update node for later update. If there is no left existing nodes in
+// `existing_nodes` to select from, generates a random makeup node under
+// `insert_type` and inserts it into db. Returns detailed error if query
+// executions failed.
 template <typename T, typename N>
-void GenerateNodes(const NodesParam& nodes_param, const T& type,
-                   google::protobuf::RepeatedPtrField<N>& nodes,
-                   int64& curr_bytes) {
+tensorflow::Status PrepareNodeForUpdate(const T& insert_type, const int64 i,
+                                        MetadataStore& store,
+                                        std::vector<Node>& existing_nodes,
+                                        N& node, T& update_type) {
+  if (!existing_nodes.empty()) {
+    Node existing_node = existing_nodes.back();
+    node = absl::get<N>(existing_node);
+    TF_RETURN_IF_ERROR(SetTypeForUpdateNode(node, store, update_type));
+    // Removes this existing update node from `existing_nodes` for avoiding
+    // abort errors under multi-thread environment.
+    existing_nodes.pop_back();
+  } else {
+    // Sets `update_type` to `insert_type` for makeup node.
+    update_type = insert_type;
+    node.set_type_id(insert_type.id());
+    node.set_name(
+        absl::StrCat("makeup_node", absl::FormatTime(absl::Now()), "_", i));
+    // Inserts the makeup node into db.
+    TF_RETURN_IF_ERROR(PrepareMakeUpNodeInDb(store, node));
+  }
+  return tensorflow::Status::OK();
+}
+
+// Sets node's properties and custom properties given `type`.
+template <typename T, typename N>
+int64 SetNodePropertiesGivenType(const FillNodesConfig& fill_nodes_config,
+                                 const NodesParam& nodes_param, const T& type,
+                                 N& node) {
+  int64 num_custom_properties_to_clear;
+  if (fill_nodes_config.update()) {
+    // Ensures that for update cases, the number of properties / custom
+    // properties being added, deleted or updated is equal to
+    // `nodes_param.num_properties`.
+    num_custom_properties_to_clear = std::min(
+        (int64)node.custom_properties_size(), nodes_param.num_properties);
+    auto it = node.custom_properties().begin();
+    for (int64 i = 0; i < num_custom_properties_to_clear; ++i) {
+      node.mutable_custom_properties()->erase(it->first);
+      it++;
+    }
+  } else {
+    // For insert cases, this will be a no-op.
+    num_custom_properties_to_clear = 0;
+  }
+
+  // If there are no properties that needed to be changed further, return
+  // directly.
+  int64 remain_num_properties_to_change =
+      nodes_param.num_properties - num_custom_properties_to_clear;
+  if (remain_num_properties_to_change == 0) {
+    return GetTransferredBytes<N>(node);
```

Comment:
should the transferred bytes be calculated differently for the update case?
we probably should count the ones that are deleted, and the properties that are changed or inserted, ignore the properties that are the same.

Reply:
Good point! Noted. I create two additional variable `transferred_bytes_for_cleared_node_properties` and `transferred_bytes_for_added_or_updated_node_properties` accordingly. PTAL.